File: bkk_response_parser.py
from typing import Optional, List, Dict, Any
from datetime import datetime
from dataclasses import dataclass
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


@dataclass
class Arrival:
    """Data class for arrival information"""
    line: str
    destination: str
    departure_time: Optional[str] = None
    departs_in_min: Optional[int] = None
    timestamp: Optional[int] = None


def find_route_id_for_trip(trip_id: str, references: Dict[str, Any]) -> Optional[str]:
    """Helper function to find route ID for a given trip ID"""
    trips = references.get("trips", {})
    trip = trips.get(trip_id)
    if trip:
        return trip.get("routeId")
    return None

def find_route_name_for_route_id(route_id: str, references: Dict[str, Any]) -> Optional[str]:
    """Helper function to find route name for a given route ID"""
    routes = references.get("routes", {})
    route = routes.get(route_id)
    if route:
        return route.get("shortName") 
    return None


def parse_arrivals_response(response: Dict[str, Any]) -> List[Arrival]:
    # fetched successfully, now parse the response: 

    # print(json.dumps(response, indent=2))  # Debug: print raw response
    arrivals = []
    
    if not response.get("data") or not response["data"].get("entry"):
        logger.info(f"No schedule info for the given stop")
        return []

    apiTime_s = response["currentTime"] / 1000  # Convert ms to seconds
    
    try:
        entry = response["data"]["entry"]
        stop_times = entry.get("stopTimes", [])
        references = response["data"].get("references", {})
    except KeyError as e:
        logger.error(f"Failed to parse entry from response: {e}")
        return []


    
    # References are dictionaries keyed by ID (from OTP dialect)
    trips = references.get("trips", {})
    
    # Parse arrivals
    for stop_time in stop_times:
        trip_id = stop_time.get("tripId")
        route_id = find_route_id_for_trip(trip_id, references)
        route_name = find_route_name_for_route_id(route_id, references)

        try: 
            depertureTime = stop_time.get("predictedArrivalTime") or stop_time.get("arrivalTime")
        except Exception as e:
            print(f"Error parsing departure time for stop_time: {stop_time}")
            logger.error(f"Failed to parse departure time: {e}")
        
        trip = trips.get(trip_id, {})

        
        departs_in_min = int(round((depertureTime - apiTime_s) / 60)) \
            if depertureTime else None

        
        # Get destination
        destination = trip.get("tripHeadsign", "?")

        depertureTime_str = datetime.fromtimestamp(depertureTime).strftime("%H:%M") 
        
        arrival = Arrival(
            line=route_name or route_id or "n.a.",
            destination=destination,
            departure_time=depertureTime_str,
            departs_in_min=departs_in_min,
            timestamp=depertureTime
        )
        arrivals.append(arrival)
    
    return arrivals

File: cpp/bkk_uds_server/bkk_response_parser.cpp
#include "bkk_response_parser.hpp"
#include <nlohmann/json.hpp>
#include <cmath>
#include <cstring>
#include <iostream>
#include <ctime>
#include <sstream>
#include <iomanip>

using json = nlohmann::json;

// parse arrival response from BKK Server 

static std::string find_route_id_for_trip(
    const std::string& trip_id, 
    const json& references) {
    
    // trips reference contains trip details    
    try {
        if (references.contains("trips") && references["trips"].contains(trip_id)) {
            auto trip = references["trips"][trip_id];
            if (trip.contains("routeId")) {
                // we need the rout id to find the line name
                return trip["routeId"].get<std::string>();
            }
        }
    } 
    catch (const std::exception& e) {
        std::cerr << "Error finding route ID for trip: " << e.what() << std::endl;
    }
    return "";
}

static std::string find_route_name_for_route_id(
    const std::string& route_id, 
    const json& references) {
    
    // the routes reference contins route info 
    try {
        if (references.contains("routes") && references["routes"].contains(route_id)) {
            auto route = references["routes"][route_id];
            if (route.contains("shortName")) {
                // short name is the line name 
                return route["shortName"].get<std::string>();
            }
        }
    } 
    catch (const std::exception& e) {
        std::cerr << "Error finding route name for route ID: " << e.what() << std::endl;
    }
    return "";
}

std::vector<Arrival> parse_arrivals_response(const std::string& response_body) {
    std::vector<Arrival> arrivals;

    json response;
    try {
        response = json::parse(response_body);
    } catch (const std::exception& e) {
        std::cerr << "Failed to parse response JSON: " << e.what() << std::endl;
        return arrivals;
    }
    
    // Check if response has required data structure
    if (!response.contains("data") || !response["data"].contains("entry")) {
        std::cout << "No schedule info for the given stop" << std::endl;
        return arrivals;
    }
    
    try {
        // Get API time in seconds (convert from milliseconds)
        double apiTime_s = response["currentTime"].get<long long>() / 1000.0;
        
        const auto& entry = response["data"]["entry"];

        // stop info: 
        const auto& stop_times = entry.contains("stopTimes") ? 
            entry["stopTimes"] : json::array();

        // references contains extra info about trip id and route details 
        const auto& references = response["data"].contains("references") ? 
            response["data"]["references"] : json::object();
        
        // Get trips reference
        const auto& trips = references.contains("trips") ? 
            references["trips"] : json::object();
        
        // Parse each stop time
        for (const auto& stop_time : stop_times) {
            try {
                // trip id of the current stop time: get line info 
                std::string trip_id = stop_time.contains("tripId") ? 
                    stop_time["tripId"].get<std::string>() : "";
                
                // find route information
                std::string route_id = find_route_id_for_trip(trip_id, references);
                std::string line = find_route_name_for_route_id(route_id, references);
                if (line.empty()) {
                    line = !route_id.empty() ? route_id : "n.a.";
                }
                
                // Get departure time 
                long long departure_time = 0;
                if (stop_time.contains("predictedArrivalTime") && 
                    stop_time["predictedArrivalTime"].is_number()) {
                    // try to get predicted: 
                    departure_time = stop_time["predictedArrivalTime"].get<long long>();
                } else if (stop_time.contains("arrivalTime") && 
                           stop_time["arrivalTime"].is_number()) {

                    // if not available get scheduled (always available) 
                    departure_time = stop_time["arrivalTime"].get<long long>();
                }
                
                // Calculate minutes until departure
                int departs_in_min = (int)std::round((departure_time - apiTime_s) / 60.0);
                
                // Format departure time as HH:MM
                std::time_t time_t_val = (std::time_t)departure_time;
                std::tm* timeinfo = std::localtime(&time_t_val);
                std::stringstream ss;
                ss << std::put_time(timeinfo, "%H:%M");
                std::string departure_time_str = ss.str();
                
                // Get destination
                std::string destination = "?";
                if (trips.contains(trip_id) && trips[trip_id].contains("tripHeadsign")) {
                    destination = trips[trip_id]["tripHeadsign"].get<std::string>();
                }
                
                // Create and add arrival
                Arrival arrival {};
                strncpy(arrival.line_id, 
                    line.c_str(), sizeof(arrival.line_id) - 1);
                strncpy(arrival.destination, 
                    destination.c_str(), sizeof(arrival.destination) - 1);
                strncpy(arrival.departure_time, 
                    departure_time_str.c_str(), sizeof(arrival.departure_time) - 1);
                arrival.departs_in_min = departs_in_min;
                arrival.timestamp = departure_time;
                
                arrivals.push_back(arrival);
                
            } catch (const std::exception& e) {
                std::cerr << "Error parsing stop time: " << e.what() << std::endl;
                continue;
            }
        }
        
    } catch (const std::exception& e) {
        std::cerr << "Failed to parse entry from response: " << e.what() << std::endl;
    }
    
    return arrivals;
}